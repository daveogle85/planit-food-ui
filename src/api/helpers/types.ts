export enum RestVerb {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  DELETE = 'DELETE',
  OPTIONS = 'OPTIONS',
}

export type Options = {
  method: RestVerb;
  redirect: 'follow' | 'error' | 'manual' | undefined;
  headers: Headers;
  body?: string;
};
