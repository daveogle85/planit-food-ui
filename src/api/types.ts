export type ApiDay = {
  date: string;
};
