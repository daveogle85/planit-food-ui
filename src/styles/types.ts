export type EmotionProps = {
  className?: string;
};
