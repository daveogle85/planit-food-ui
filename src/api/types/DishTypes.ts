import { Ingredient } from './IngredientsTypes';

export enum DishType {
  MAIN,
  SIDE,
  DESSERT,
}

export type ApiDish = {
  id?: string;
  name?: string;
  searchName?: string;
  dishType?: DishType;
  notes?: string;
  cookingTime?: number;
  ingredients?: Array<Ingredient>;
};

export type Dish = ApiDish & {
  localId: string;
};
