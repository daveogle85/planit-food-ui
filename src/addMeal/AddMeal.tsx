import classNames from 'classnames';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { searchForDish } from '../api/dish';
import { searchForMeal } from '../api/meal';
import { ApiDish, Dish, DishType } from '../api/types/DishTypes';
import { ApiMeal, Meal } from '../api/types/MealTypes';
import AutoCompleteInput from '../components/AutoCompleteInput/AutoCompleteInput';
import FeedbackElement from '../components/FeedbackInput/FeedbackElement';
import NavBar from '../components/NavBar/NavBar';
import { authSelectors } from '../ducks/auth/AuthReducer';
import { ToastState, FeedbackStatus } from '../ducks/toast/ToastTypes';
import {
  addMealToSelectedList,
  fetchLists,
  listsSelectors,
} from '../ducks/lists/ListsReducer';
import { nullOrEmptyString } from '../helpers/string';
import { EmotionProps } from '../styles/types';
import { DishComponentProps } from './AddMealTypes';
import { styledAddMeal } from './StyledAddMeal';
import Loading from '../components/Loading/Loading';

function generateMealName(dishes: Array<Dish>): string {
  if (!dishes.length) {
    return '';
  }

  const sortedDishes = dishes.sort((a, b) =>
    a.dishType === DishType.MAIN && b.dishType === DishType.SIDE ? -1 : 1
  );

  return sortedDishes
    .map(s => s.name)
    .join(', ')
    .replace(/,([^,]*)$/, ' &$1');
}

const AddMeal: React.FC<EmotionProps> = props => {
  const defaultDish: Dish = {
    localId: 'local-0',
    name: '',
    dishType: DishType.MAIN,
  };
  const defaultDishes = [defaultDish];
  const defaultMeal: Meal = {
    localId: 'local-meal-0',
    name: '',
  };

  const token = useSelector(authSelectors.selectedToken);
  const dispatch = useDispatch();
  const [dishCount, setDishCount] = useState(0);
  const [dishes, setDishes] = useState([defaultDish]);
  const [meal, setMeal] = useState<Meal>(defaultMeal);
  const [isMealDirty, setIsMealDirty] = useState(false);
  const loadingLists = useSelector(listsSelectors.selectLoading);
  const selectedList = useSelector(listsSelectors.selectSelectedList);
  const [dishErrors, setDishErrors] = useState(new Map<string, string>());

  useEffect(() => {
    async function getLists() {
      dispatch(fetchLists());
    }
    getLists();
  }, [dispatch, token]);

  const getMealOptions = (text: string) => searchForMeal(text)(token);

  const getDishId = () => {
    const id = `local-${dishCount + 1}`;
    setDishCount(dishCount + 1);
    return id;
  };

  const setDishesAndMealName = (dishes: Array<Dish>) => {
    const namedDishes = dishes.filter(d => !nullOrEmptyString(d.name));
    if (!isMealDirty) {
      setMeal({
        ...meal,
        name: generateMealName(namedDishes),
      });
    }
    setDishes(dishes);
  };

  const handleDeleteDish = (dishId: string) => {
    const newDishes = dishes.filter(dish => dish.localId !== dishId);
    setDishesAndMealName(newDishes);
  };

  const handleAddDish = () => {
    const newDishes = [...dishes];
    newDishes.push({
      localId: getDishId(),
      name: '',
      dishType: dishes.length === 0 ? DishType.MAIN : DishType.SIDE,
    });
    setDishesAndMealName(newDishes);
  };

  const handleDishUpdate = (oldId: string) => (dishToUpdate: Dish) => {
    let nameCount = 0;
    updateDishErrors();
    const newDishes = dishes.map(dish => {
      if (nameCount > 0) {
        const errors = new Map(dishErrors);
        errors.set(dishToUpdate.localId, 'Dishes must have unique names');
        setDishErrors(errors);
      }

      if (dish.name === dishToUpdate.name) {
        nameCount++;
      }

      if (dish.localId === oldId) {
        return dishToUpdate;
      }
      return dish;
    });
    setDishesAndMealName(newDishes);
  };

  const handleMealUpdate = (mealToUpdate: Meal) => {
    let dishesToUpdate: Array<Dish> = [];
    const inDatabase = !nullOrEmptyString(mealToUpdate.id);

    if (inDatabase) {
      dishesToUpdate = mealToUpdate.dishes ?? [];
    } else {
      // If meal not in database clear out the dishes if we have just
      // switched from a db meal to a custom
      dishesToUpdate =
        !nullOrEmptyString(meal.id) && !inDatabase
          ? defaultDishes
          : dishes ?? defaultDishes;
    }

    setMeal(mealToUpdate);
    setDishes(dishesToUpdate);
  };

  const handleMainChecked = (id: string, checked: boolean) => {
    const newDishes = dishes.map(dish =>
      dish.localId === id
        ? { ...dish, dishType: checked ? DishType.MAIN : DishType.SIDE }
        : dish
    );
    setDishesAndMealName([...newDishes]);
  };

  const handleMealClear = () => {
    setMeal(defaultMeal);
    setDishesAndMealName(defaultDishes);
    setIsMealDirty(false);
  };

  const handleAddMealToList = () => {
    const dishesToAdd = dishes.filter(d => !nullOrEmptyString(d.name));
    const mealToAdd = { ...meal, dishes: dishesToAdd };
    dispatch(addMealToSelectedList(mealToAdd));
  };

  const convertFromDishApi = (dish: ApiDish, id?: string) => ({
    ...dish,
    localId: id ?? getDishId(),
  });

  const convertFromMealApi = (meal: ApiMeal) => {
    let id = dishCount;
    const dishes = meal.dishes?.map(d => {
      const dish = convertFromDishApi(d, `local-${id + 1}`);
      id++;
      return dish;
    });
    setDishCount(dishCount + (meal.dishes?.length ?? 0));
    return { ...meal, localId: 'local-meal-1', dishes };
  };

  function updateDishErrors() {
    const errorsToRemove: Array<string> = [];
    const errors = new Map(dishErrors);
    errors.forEach((_, key) => {
      const duplicates = dishes.filter(dish => dish.localId === key);
      if (duplicates.length <= 1) {
        errorsToRemove.push(key);
      }
    });
    errorsToRemove.forEach(err => errors.delete(err));
    setDishErrors(errors);
  }

  const DishComponent = (props: DishComponentProps) => {
    const getDishOptions = (text: string) => searchForDish(text)(token);
    const { dish } = props;
    return (
      <>
        <AutoCompleteInput
          className="dish-name"
          placeholder="Dish Name"
          getOptions={getDishOptions}
          currentValue={dish}
          updateCurrentValue={handleDishUpdate(dish.localId)}
          inputError={dishErrors.get(dish.localId)}
          disabled={!nullOrEmptyString(meal.id)}
          convertFromApiType={convertFromDishApi}
        />
        <div>
          <input
            name="isMain"
            type="checkbox"
            checked={props.dish.dishType === DishType.MAIN}
            onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
              handleMainChecked(dish.localId, e.currentTarget.checked)
            }
            disabled={!nullOrEmptyString(meal.id)}
          />
        </div>
        <div className="delete-dish">
          <button
            disabled={!nullOrEmptyString(meal.id)}
            onClick={() => handleDeleteDish(props.dish.localId)}
          >
            -
          </button>
        </div>
      </>
    );
  };

  const getSelectedListText = () =>
    selectedList == null ? 'No List Available' : selectedList.name;

  const validateMeal = (): ToastState => {
    let message = undefined;
    let status = FeedbackStatus.HIDDEN;

    if (selectedList == null) {
      status = FeedbackStatus.ERROR;
      message = 'No Default list was found';
    }

    // no dish names and no meal name
    const atLeastOneDishName = dishes.reduce(
      (a, c) => (a = a || !nullOrEmptyString(c.name)),
      false
    );
    const mealHasName = !nullOrEmptyString(meal.name);

    if (!atLeastOneDishName && !mealHasName) {
      status = FeedbackStatus.ERROR;
      message = 'Meals with no dishes must have a name';
    }

    if (dishErrors.size > 0) {
      status = FeedbackStatus.ERROR;
      message = 'Error in dishes';
    }

    return {
      status,
      message,
    };
  };

  const dishesDisabledState = () => {
    if (!nullOrEmptyString(meal.id)) {
      return {
        status: FeedbackStatus.DISABLED,
        message: 'Dishes associated with a meal cannot be edited',
      };
    }
    return {
      status: FeedbackStatus.HIDDEN,
    };
  };

  return (
    <>
      <NavBar />
      <Loading isLoading={loadingLists}>
        <div className={classNames('AddMeal', props.className)}>
          <h2 className="title">Add A New Meal</h2>
          <div>Selected List: {getSelectedListText()}</div>
          <div>
            <div className="meal">
              <label htmlFor="mealname">Meal name</label>
              <div>
                <AutoCompleteInput
                  className="meal-name"
                  placeholder="Optional"
                  getOptions={getMealOptions}
                  updateCurrentValue={handleMealUpdate}
                  currentValue={meal}
                  onDirty={setIsMealDirty}
                  convertFromApiType={convertFromMealApi}
                />
                <button onClick={handleMealClear}>Clear Meal</button>
              </div>
            </div>
            <FeedbackElement
              className="dishes-wrapper"
              state={dishesDisabledState()}
            >
              <div className="dishes">
                <h3>Dishes</h3>
                <div className="dishes-grid">
                  <div>Dish Name</div>
                  <div>Main</div>
                  <div>Delete</div>
                  {dishes.map(dish => (
                    <DishComponent key={dish.localId} dish={dish} />
                  ))}
                </div>
                <button
                  disabled={!nullOrEmptyString(meal.id)}
                  onClick={handleAddDish}
                >
                  Add Dish
                </button>
              </div>
            </FeedbackElement>
            <div className="options">
              <h3>Notes</h3>
              <textarea />
            </div>
            <div>
              <FeedbackElement state={validateMeal()}>
                <button
                  disabled={validateMeal().status === FeedbackStatus.ERROR}
                  onClick={handleAddMealToList}
                >
                  Add Meal To List
                </button>
              </FeedbackElement>
              <button disabled={true}>Add Meal To Calendar</button>
            </div>
          </div>
        </div>
      </Loading>
    </>
  );
};

export { generateMealName as _test_generateMealName };

export default styledAddMeal(AddMeal);
