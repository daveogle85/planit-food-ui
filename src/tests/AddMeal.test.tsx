import serializer from 'jest-emotion';
import React from 'react';

import AddMeal, { _test_generateMealName } from '../views/addMeal/AddMeal';
import { renderWithTheme } from './helpers';
import { RootState } from '../ducks';
import { styledAddMeal } from '../views/addMeal/StyledAddMeal';
import { DishType } from '../api/types/DishTypes';

jest.mock('../contexts/auth0-context', () => ({
  useAuth0: jest.fn().mockReturnValue({
    loginWithRedirect: jest.fn(),
    loading: false,
    isAuthenticated: false,
    user: {
      picture: '',
    },
  }),
}));

expect.addSnapshotSerializer(serializer);
test('Add Meal View renders correctly', () => {
  const mockStore: RootState = {
    lists: {
      loading: false,
      selectedList: null,
      lists: [],
    },
    auth: { isAuthenticated: true, token: '' },
  };
  const TestAddMealView = styledAddMeal(AddMeal as any);
  const view = renderWithTheme(<TestAddMealView />, mockStore).toJSON();
  expect(view).toMatchSnapshot();
});

describe('generateMealName', () => {
  it('should return an empty string if no dish names', () => {
    const result = _test_generateMealName([]);
    expect(result).toBe('');
  });

  it('should return a single dish', () => {
    const dish = [{ localId: 'test1', name: 'test1', dishType: DishType.SIDE }];
    const result = _test_generateMealName(dish);
    expect(result).toBe('test1');
  });

  it('should sort dishes by main first', () => {
    const dishes = [
      { localId: 'test1', name: 'test1', dishType: DishType.SIDE },
      { localId: 'test2', name: 'test2', dishType: DishType.MAIN },
      { localId: 'test3', name: 'test3', dishType: DishType.SIDE },
      { localId: 'test4', name: 'test4', dishType: DishType.SIDE },
      { localId: 'test5', name: 'test5', dishType: DishType.MAIN },
    ];
    const result = _test_generateMealName(dishes);
    expect(result).toBe('test2, test5, test1, test3 & test4');
  });
});
