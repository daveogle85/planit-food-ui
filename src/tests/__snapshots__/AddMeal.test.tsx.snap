// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Meal View renders correctly 1`] = `
Array [
  .emotion-4 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  top: 0;
  left: 0;
  right: 0;
  padding: 0;
  height: 70px;
  box-shadow: 4px 0 12px -6px #000, -4px 0 12px -6px #000;
  list-style-type: none;
  z-index: 2;
}

.emotion-4 ul {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  width: 100%;
  margin: 0;
  padding-left: 40px;
}

.emotion-4 ul li {
  cursor: pointer;
  list-style-type: none;
  height: 100%;
  min-width: 30px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-4 ul li:hover:not(.app-title) {
  background-color: #f1f1f1;
}

.emotion-4 ul .app-title {
  white-space: nowrap;
}

.emotion-4 ul .dd-menu-items {
  position: relative;
  height: calc(100% - 0px);
  width: calc(100% - 0px);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.emotion-4 ul .dd-menu-items li:not(.dd-list-item) {
  margin-right: 8px;
}

.emotion-1 {
  height: 100%;
  position: relative;
}

.emotion-1 .dd-header {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

.emotion-1 .dd-header-title {
  cursor: pointer;
  padding: 0 8px;
}

.emotion-1 li:hover {
  background-color: #46627f;
}

.emotion-1 .dd-list {
  z-index: 1;
  background-color: #fff;
  position: absolute;
  padding: 0;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  display: none;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  min-width: 100%;
  border: 1px solid #2b2d2f;
  border-radius: 0.3em;
}

.emotion-1 .dd-list.dd-open {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 .dd-list.dd-oob {
  right: 0;
}

.emotion-1 .dd-list li {
  cursor: pointer;
  border-radius: 0.3em;
  padding: 0 8px;
  width: -webkit-fill-available;
  width: -moz-available;
  width: fill-available;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 30px;
  line-height: 30px;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0 8px;
}

.emotion-0 .add-text {
  display: none;
}

.emotion-0 svg {
  position: relative;
  display: inline-block;
  border: 2px solid #9b9797;
  width: 40px;
  height: 40px;
  border-radius: 2em;
}

@media (min-width:768px) {
  .emotion-0 svg {
    margin-right: 8px;
    width: 50px;
    height: 50px;
  }
}

@media (min-width:768px) {
  .emotion-0 .add-text {
    display: block;
  }
}

.emotion-2 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 8px;
}

.emotion-2 div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 .details {
  display: none;
  font-weight: 300;
  font-size: 15px;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media (min-width:768px) {
  .emotion-2 .details {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

.emotion-2 img {
  width: 40px;
  height: 40px;
  border-radius: 2em;
}

@media (min-width:768px) {
  .emotion-2 img {
    margin-right: 8px;
    width: 50px;
    height: 50px;
  }
}

<nav
    className="emotion-4"
  >
    <ul
      className="menu"
    >
      <li
        className="app-title"
        onClick={[Function]}
      >
        Planit Food App
      </li>
      <div
        className="dd-menu-items"
      >
        <li
          className="dd-add"
        >
          <div
            className="dd-wrapper emotion-1"
          >
            <div
              className="dd-header"
              onClick={[Function]}
            >
              <div
                className="emotion-0"
                title="Add Options"
              >
                <svg
                  id="svg1"
                  version="1.0"
                  viewBox="0 0 60 60"
                  x="0"
                  y="0"
                >
                  <path
                    d="m23.78 23.78v-13.999c0-1.9588 0.536-3.55 1.607-4.7741 1.071-1.2546 2.601-1.8819 4.59-1.8819s3.519 0.6273 4.59 1.8819c1.102 1.2241 1.653 2.8153 1.653 4.7737v13.999h13.816c2.019 0 3.611 0.551 4.773 1.653 1.194 1.071 1.791 2.586 1.791 4.544 0 1.989-0.597 3.534-1.791 4.636-1.162 1.102-2.754 1.652-4.773 1.652h-13.816v13.954c0 1.989-0.551 3.596-1.653 4.82-1.101 1.224-2.632 1.836-4.59 1.836s-3.488-0.612-4.59-1.836c-1.071-1.224-1.607-2.831-1.607-4.82v-13.954h-13.816c-1.9582 0-3.5494-0.596-4.7735-1.79-1.1934-1.224-1.7901-2.723-1.7901-4.498 0-1.958 0.5814-3.473 1.7442-4.544 1.1935-1.102 2.8-1.653 4.8196-1.653h13.816z"
                    fill="none"
                    id="path1092"
                    stroke="#ffffff"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth={5}
                  />
                  <path
                    d="m23.78 23.78v-13.999c0-1.9588 0.536-3.55 1.607-4.7741 1.071-1.2546 2.601-1.8819 4.59-1.8819s3.519 0.6273 4.59 1.8819c1.102 1.2241 1.653 2.8153 1.653 4.7737v13.999h13.816c2.019 0 3.611 0.551 4.773 1.653 1.194 1.071 1.791 2.586 1.791 4.544 0 1.989-0.597 3.534-1.791 4.636-1.162 1.102-2.754 1.652-4.773 1.652h-13.816v13.954c0 1.989-0.551 3.596-1.653 4.82-1.101 1.224-2.632 1.836-4.59 1.836s-3.488-0.612-4.59-1.836c-1.071-1.224-1.607-2.831-1.607-4.82v-13.954h-13.816c-1.9582 0-3.5494-0.596-4.7735-1.79-1.1934-1.224-1.7901-2.723-1.7901-4.498 0-1.958 0.5814-3.473 1.7442-4.544 1.1935-1.102 2.8-1.653 4.8196-1.653h13.816z"
                    fill="#9b9797"
                    id="path837"
                  />
                </svg>
                <div
                  className="add-text"
                >
                  Add
                </div>
              </div>
            </div>
            <div>
              <ul
                className="dd-list"
              >
                <li
                  className="dd-list-item"
                  onClick={[Function]}
                >
                  Add A Meal
                </li>
              </ul>
            </div>
          </div>
        </li>
        <li
          className="dd-profile"
        >
          <div
            className="dd-wrapper emotion-1"
          >
            <div
              className="dd-header"
              onClick={[Function]}
            >
              <div
                className="emotion-2"
                title="Logout"
              >
                <img
                  alt="Avatar"
                  src=""
                />
                <div
                  className="details"
                >
                  <span />
                  <span />
                </div>
              </div>
            </div>
            <div>
              <ul
                className="dd-list"
              >
                <li
                  className="dd-list-item"
                  onClick={[Function]}
                >
                  Logout
                </li>
              </ul>
            </div>
          </div>
        </li>
      </div>
    </ul>
  </nav>,
  .emotion-0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  right: 0;
  padding: 0;
  border: 1px solid #000;
  border-right: none;
  border-top: none;
  border-bottom-left-radius: 2em;
  background-color: #7f53ac;
  position: absolute;
  top: 70px;
  bottom: 0;
  width: 0;
  max-width: calc(100% - 40px);
  z-index: 2;
  -webkit-transition: 0.5s;
  transition: 0.5s;
}

.emotion-0 .tab {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: absolute;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  left: -41px;
  top: 0;
  width: 40px;
  background-color: #7f53ac;
}

.emotion-0 .tab > div {
  position: relative;
  padding: 10px;
  writing-mode: sideways-lr;
}

.emotion-0 .tab .spinner {
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.emotion-0 .tab:after {
  content: '';
  position: absolute;
  z-index: 1;
  background-color: #7f53ac;
  box-shadow: 0 2px 2px rgba(0,0,0,0.4);
  -webkit-transform: skewY(10deg);
  -ms-transform: skewY(10deg);
  transform: skewY(10deg);
  border-radius: 0 0 0 5px;
  height: 1em;
  width: 100%;
  bottom: -0.5em;
}

.emotion-0 ul {
  padding: 0;
  list-style: none;
}

.emotion-0:not(.draw-open) .draw-content {
  display: none;
}

.emotion-0 .draw-content {
  padding: 8px;
}

.emotion-0.draw-open {
  width: 400px;
}

<div
    className="list emotion-0"
  >
    <div
      className="tab"
      onClick={[Function]}
    >
      <div>
        
      </div>
    </div>
  </div>,
  .emotion-6 {
  margin: 30px;
  margin: 30px;
}

.emotion-6 .title {
  font-weight: 700;
  font-size: 20px;
}

.emotion-6 label {
  padding-right: 8px;
}

.emotion-6 .meal {
  padding-bottom: 8px;
}

.emotion-6 .dishes {
  padding: 8px;
}

.emotion-6 .dishes-wrapper .feedback-icon {
  top: 8px;
}

.emotion-6 .dishes-wrapper .feedback-icon:after {
  right: -25px;
}

.emotion-6 button {
  padding-top: 4px;
  padding-bottom: 4px;
}

.emotion-6 .options {
  padding-bottom: 20px;
}

.emotion-6 textarea {
  width: 100%;
}

.emotion-6 .dishes,
.emotion-6 .dishes-wrapper {
  width: 100%;
}

.emotion-6 .dishes-grid {
  display: grid;
  grid-template-columns: 4fr 1fr 1fr;
  grid-auto-rows: 30px;
  grid-gap: 8px;
  margin-bottom: 20px;
  padding-bottom: 8px;
}

.emotion-6 .dishes-grid .delete-dish > button {
  height: 100%;
  width: 30px;
}

.emotion-6 .title {
  font-weight: 700;
  font-size: 20px;
}

.emotion-6 label {
  padding-right: 8px;
}

.emotion-6 .meal {
  padding-bottom: 8px;
}

.emotion-6 .dishes {
  padding: 8px;
}

.emotion-6 .dishes-wrapper .feedback-icon {
  top: 8px;
}

.emotion-6 .dishes-wrapper .feedback-icon:after {
  right: -25px;
}

.emotion-6 button {
  padding-top: 4px;
  padding-bottom: 4px;
}

.emotion-6 .options {
  padding-bottom: 20px;
}

.emotion-6 textarea {
  width: 100%;
}

.emotion-6 .dishes,
.emotion-6 .dishes-wrapper {
  width: 100%;
}

.emotion-6 .dishes-grid {
  display: grid;
  grid-template-columns: 4fr 1fr 1fr;
  grid-auto-rows: 30px;
  grid-gap: 8px;
  margin-bottom: 20px;
  padding-bottom: 8px;
}

.emotion-6 .dishes-grid .delete-dish > button {
  height: 100%;
  width: 30px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  max-width: 500px;
  height: 30px;
}

.emotion-1 .dd-list {
  z-index: 1;
  background-color: #fff;
  position: absolute;
  padding: 0;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  display: none;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  min-width: 100%;
  border: 1px solid #2b2d2f;
  border-radius: 0.3em;
}

.emotion-1 .dd-list.dd-open {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 .dd-list.dd-oob {
  right: 0;
}

.emotion-1 .dd-list li {
  cursor: pointer;
  border-radius: 0.3em;
  padding: 0 8px;
  width: -webkit-fill-available;
  width: -moz-available;
  width: fill-available;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 30px;
  line-height: 30px;
}

.emotion-1 input {
  width: 100%;
  height: 100%;
}

.emotion-1 .feedback-hidden > input {
  border: 1px solid #f1f1f1;
  border-radius: 0.3em;
  height: calc(100% - 6px);
}

.emotion-1 > div {
  width: 100%;
}

.emotion-1 .dd-list {
  margin: 0;
  top: 30px;
}

.emotion-1 .dd-list li:hover {
  background-color: #46627f;
}

.emotion-0 {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0:not(.feedback-hidden) > *:first-child {
  padding-right: 25px;
  border: 2px solid #9b9797;
  border-radius: 0.3em;
  height: calc(100% - 6px);
}

.emotion-0 .feedback-icon {
  width: 13px;
  height: 13px;
  position: absolute;
  right: 5px;
  font-size: 12px;
  font-weight: 700;
  color: #9b9797;
  border-radius: 2em;
  border: 2px solid #9b9797;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
}

.emotion-0 .feedback-icon:after {
  content: '';
  display: none;
  position: absolute;
  white-space: nowrap;
  border: 1px solid #9b9797;
  border-radius: 0.3em;
  padding: 2px;
  bottom: 150%;
  background: white;
  opacity: 0.9;
  font-size: 8px;
}

.emotion-0 .feedback-icon:hover:after {
  display: block;
}

.emotion-5 {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-5:not(.feedback-hidden) > *:first-child {
  padding-right: 25px;
  border: 2px solid #cc0000;
  border-radius: 0.3em;
  height: calc(100% - 6px);
}

.emotion-5 .feedback-icon {
  width: 13px;
  height: 13px;
  position: absolute;
  right: 5px;
  font-size: 12px;
  font-weight: 700;
  color: #cc0000;
  border-radius: 2em;
  border: 2px solid #cc0000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
}

.emotion-5 .feedback-icon:after {
  content: 'Meals with no dishes must have a name';
  display: none;
  position: absolute;
  white-space: nowrap;
  border: 1px solid #cc0000;
  border-radius: 0.3em;
  padding: 2px;
  bottom: 150%;
  background: white;
  opacity: 0.9;
  font-size: 8px;
}

.emotion-5 .feedback-icon:hover:after {
  display: block;
}

<div
    className="AddMeal emotion-6"
  >
    <h2
      className="title"
    >
      Add A New Meal
    </h2>
    <div>
      <div
        className="meal"
      >
        <label
          htmlFor="mealname"
        >
          Meal name
        </label>
        <div>
          <div
            className="auto-complete-input meal-name emotion-1"
          >
            <div
              className="feedback-hidden emotion-0"
            >
              <input
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Optional"
                type="text"
                value=""
              />
            </div>
            <ul
              className="dd-list"
            >
              <li>
                No Option Found
              </li>
            </ul>
          </div>
          <button
            onClick={[Function]}
          >
            Clear Meal
          </button>
        </div>
      </div>
      <div
        className="dishes-wrapper feedback-hidden emotion-0"
      >
        <div
          className="dishes"
        >
          <h3>
            Dishes
          </h3>
          <div
            className="dishes-grid"
          >
            <div>
              Dish Name
            </div>
            <div>
              Main
            </div>
            <div>
              Delete
            </div>
            <div
              className="auto-complete-input dish-name emotion-1"
            >
              <div
                className="feedback-hidden emotion-0"
              >
                <input
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Dish Name"
                  type="text"
                  value=""
                />
              </div>
              <ul
                className="dd-list"
              >
                <li>
                  No Option Found
                </li>
              </ul>
            </div>
            <div>
              <input
                checked={true}
                disabled={false}
                name="isMain"
                onChange={[Function]}
                type="checkbox"
              />
            </div>
            <div
              className="delete-dish"
            >
              <button
                disabled={false}
                onClick={[Function]}
              >
                -
              </button>
            </div>
          </div>
          <button
            disabled={false}
            onClick={[Function]}
          >
            Add Dish
          </button>
        </div>
      </div>
      <div
        className="options"
      >
        <h3>
          Notes
        </h3>
        <textarea />
      </div>
      <div>
        <div
          className=" emotion-5"
        >
          <button
            disabled={true}
            onClick={[Function]}
          >
            Add Meal To List
          </button>
          <div
            className="feedback-icon"
          >
            i
          </div>
        </div>
        <button
          disabled={true}
        >
          Add Meal To Calendar
        </button>
      </div>
    </div>
  </div>,
]
`;
